GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -18,6 +18,19 @@ public class NamedPipeStreamReader
         private byte[] buffer;
         private Stream stream;
 
+        /// <summary>
+        /// The index in the buffer that the next message starts at. If this
+        /// is greater than or equal to the number of bytes that have been read into the buffer,
+        /// then we need to read more bytes from the stream.
+        /// </summary>
+        private int nextMessageChunkStartIndex;
+
+        /// <summary>
+        /// The number of bytes that have been read from the stream and into
+        /// the buffer.
+        /// </summary>
+        private int numBytesReadIntoBuffer;

```

Comment:
Or yet another option would be to get rid of buffer altogether, and only read one byte at a time until you find the message terminator. How much would that affect perf? If it's not a significant hit, that will eliminate most of the complexity here.

Reply:
> else just make the buffer big enough that it won't need to wrap and simplify this down

The message can contain user entered text (git commands) and arbitrary content (Modified Paths entries) - I don't think we can put a cap on the message buffer.

> another option would be to get rid of buffer altogether,

I have not profiled it, but the msdn documentation includes a note about how inefficient this approach is [Notes to Inheritors](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.readbyte?view=netframework-4.7.2#notes-to-inheritors).

Reply:
There's no question that reading one byte at a time will be slower, but how much will that impact our actual perf numbers on git commands?

If that doesn't pan out because the perf is bad, then let's chat about options for making this solution simpler.